utils/yoke/src/yoke.rs

```diff
@@ -480,50 +471,183 @@ where
+    /// #     {
+    /// #         unsafe { f(mem::transmute(self)) }
+    /// #     }
+    /// # }
+    /// ```
+    //
+    // Safety docs can be found below on `__project_safety_docs()`
+    pub fn project<'this, P>(
+        &'this self,
+        f: for<'a> fn(&'this <Y as Yokeable<'a>>::Output, &'a ()) -> <P as Yokeable<'a>>::Output,
+    ) -> Yoke<P, C>
+    where
+        P: for<'a> Yokeable<'a>,
+    {
+        let p = f(self.get(), &());
         Yoke {
-            yokeable: unsafe { Y::make(self.get().clone()) },
+            yokeable: unsafe { P::make(p) },
             cart: self.cart.clone(),
         }
     }
-}
 
-impl<Y: for<'a> Yokeable<'a>, T: ?Sized> Clone for Yoke<Y, Option<Arc<T>>>
-where
-    for<'a> <Y as Yokeable<'a>>::Output: Clone,
-{
-    fn clone(&self) -> Self {
+    /// This is similar to [`Yoke::project`], however it works around it not being able to
+    /// use `FnOnce` by using an explicit capture input, until [compiler bug #84937](https://github.com/rust-lang/rust/issues/84937)
+    /// is fixed.
+    ///
+    /// See the docs of [`Yoke::project`] for how this works.
+    pub fn project_with_capture<'this, P, T>(
+        &'this self,
+        capture: T,
+        f: for<'a> fn(
+            &'this <Y as Yokeable<'a>>::Output,
+            capture: T,
+            &'a (),
+        ) -> <P as Yokeable<'a>>::Output,
+    ) -> Yoke<P, C>
+    where
+        P: for<'a> Yokeable<'a>,
+    {
+        let p = f(self.get(), capture, &());
         Yoke {
-            yokeable: unsafe { Y::make(self.get().clone()) },
+            yokeable: unsafe { P::make(p) },
             cart: self.cart.clone(),
         }
     }
-}
-
-impl<'b, Y: for<'a> Yokeable<'a>, T: ?Sized> Clone for Yoke<Y, Option<&'b T>>
-where
-    for<'a> <Y as Yokeable<'a>>::Output: Clone,
-{
-    fn clone(&self) -> Self {
-        Yoke {
-            yokeable: unsafe { Y::make(self.get().clone()) },
-            cart: self.cart,
-        }
-    }
-}
 
-impl<Y: for<'a> Yokeable<'a>> Clone for Yoke<Y, ()>
-where
-    for<'a> <Y as Yokeable<'a>>::Output: Clone,
-{
-    fn clone(&self) -> Self {
-        Yoke {
-            yokeable: unsafe { Y::make(self.get().clone()) },
-            cart: (),
-        }
-    }
+    /// (Docs are on a private function to allow the use of compile_fail doctests)
+    ///
+    /// This is safe to perform because of the choice of lifetimes on `f`, that is,
+    /// `for<'a> fn(&'this <Y as Yokeable<'a>>::Output, &'a ()) -> <P as Yokeable<'a>>::Output`.
+    ///
+    /// What we want this function to do is take a Yokeable (`Y`) that is borrowing from the cart, and
+    /// produce another Yokeable (`P`) that also borrows from the same cart. There are a couple potential
+    /// hazards here:
+    ///
+    /// - `P` ends up borrowing data from `Y` (or elsewhere) that did _not_ come from the cart,
+    ///   for example `P` could borrow owned data from a `Cow`.
+    /// - Borrowed data from `Y` escapes with the wrong lifetime
```

Comment:
Is this list complete? I'm struggling to convince myself it is. Here's a proof sketch that feels right but which I'm not willing to bet on:

If `Yoke<Y>` is destroyed, it must not invalidate `Yoke<P>`. The yoked data will still be alive, because of `CloneableCart`, so it remains to check that we don't keep any of the underlying bookkeeping data that Y would be entitled to destroy. We can't hold onto any data that lives only for `'y` (the lifetime of the `Y` object) so data owned by `Y` is out. So we can only *copy* data from `Y`, so it remains to check that copying non-`'static` data is ok, which reduces to checking if copying references is ok.

If the reference has the special `for<'a>` yoke lifetime, that means it's attached to the cart, so it can be copied, since it's the same yoke lifetime on both sides. If the reference has some other lifetime, like if we were converting `Y<'static, 'b> -> P<'static>`, then we can only copy it if `P` can express `'b` separately from the yoke lifetime, since the borrowed content is *not* yoked.

Now, the part I'm not sure on is that I've referred to "the" yoke lifetime, but one could conceive of multiple yoke lifetimes being active simultaneously? Like can you wind up in a situation where you have something silly like
```rust
struct NestedYokeable<'y> {
  inner: Box<Yoke<NestedYokeable<'y>, Cart>>,
  ref: &'y i32,
}
```
Could you sneak yoked content intended for the outer yoke into the inner yoke, thus breaking the guarantee we otherwise get from CloneableCart? Can a value of this type even be constructed safely?

Reply:
If you sneak in other lifetimes they're like adding more existential parameters, essentially.

Note that the `Yokeable` used in a `Yoke` must be `'static`, you can only sneak in other lifetimes via the cart. Yokes only have one "the" yoke lifetime because of Yokeable, all other lifetimes are in the cart and would constrain both sides in an acceptable way.

Reply:
I guess you could do

```rust
y.project(|y, _| y.inner.as_ref().clone())
```
but that's fine, since the lifetimes match up anyways (since you've cloned).

I think I'm pretty convinced this is correct. Huzzah.